File: lib/lambda-component.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';
import { EnvironmentConfig, ResourceTags } from './types';

export interface LambdaComponentArgs {
  config: EnvironmentConfig;
  tags: ResourceTags;
  environmentSuffix: string;
  vpcId: pulumi.Input<string>;
  privateSubnetIds: pulumi.Input<string[]>;
  dbEndpoint: pulumi.Input<string>;
}

/**
 * Lambda Component for payment processing
 */
export class LambdaComponent extends pulumi.ComponentResource {
  public readonly lambdaFunction: aws.lambda.Function;
  public readonly securityGroup: aws.ec2.SecurityGroup;
  public readonly role: aws.iam.Role;

  constructor(
    name: string,
    args: LambdaComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:compute:LambdaComponent', name, {}, opts);

    const {
      config,
      tags,
      environmentSuffix,
      vpcId,
      privateSubnetIds,
      dbEndpoint,
    } = args;

    // Create security group for Lambda
    this.securityGroup = new aws.ec2.SecurityGroup(
      `lambda-sg-${environmentSuffix}`,
      {
        vpcId: vpcId,
        description: `Security group for Lambda in ${config.environment}`,
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound',
          },
        ],
        tags: {
          ...tags,
          Name: `lambda-sg-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create IAM role for Lambda
    this.role = new aws.iam.Role(
      `lambda-role-${environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'lambda.amazonaws.com' },
              Action: 'sts:AssumeRole',
            },
          ],
        }),
        tags: {
          ...tags,
          Name: `lambda-role-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Attach basic Lambda execution policy
    new aws.iam.RolePolicyAttachment(
      `lambda-basic-${environmentSuffix}`,
      {
        role: this.role.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
      },
      { parent: this }
    );

    // Attach VPC execution policy
    new aws.iam.RolePolicyAttachment(
      `lambda-vpc-${environmentSuffix}`,
      {
        role: this.role.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole',
      },
      { parent: this }
    );

    // Create Lambda function for payment processing
    this.lambdaFunction = new aws.lambda.Function(
      `payment-processor-${environmentSuffix}`,
      {
        name: `payment-processor-${environmentSuffix}`,
        runtime: 'nodejs18.x',
        handler: 'index.handler',
        role: this.role.arn,
        timeout: 30,
        memorySize: 256,
        code: new pulumi.asset.AssetArchive({
          'index.js': new pulumi.asset.StringAsset(`
exports.handler = async (event) => {
  console.log('Processing payment in environment: ${config.environment}');
  console.log('Event:', JSON.stringify(event, null, 2));

  // Payment processing logic would go here
  const payment = {
    transactionId: Date.now().toString(),
    amount: event.amount || 0,
    currency: event.currency || 'USD',
    status: 'processed',
    environment: '${config.environment}',
    timestamp: new Date().toISOString(),
  };

  return {
    statusCode: 200,
    body: JSON.stringify(payment),
  };
};
          `),
        }),
        environment: {
          variables: {
            ENVIRONMENT: config.environment,
            DB_ENDPOINT: dbEndpoint,
            REGION: aws.config.region || 'us-east-1',
          },
        },
        vpcConfig: {
          subnetIds: privateSubnetIds,
          securityGroupIds: [this.securityGroup.id],
        },
        tags: {
          ...tags,
          Name: `payment-processor-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    this.registerOutputs({
      functionArn: this.lambdaFunction.arn,
      functionName: this.lambdaFunction.name,
    });
  }
}

File: archive/pulumi-py/Pr7062/lib/lambda_component.py
"""
Lambda Component for Payment Processing.
Creates Lambda functions with environment-specific configurations.
"""

from typing import Optional, Dict
import json
import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions, AssetArchive, FileArchive
from lib.environment_config import EnvironmentConfig


class LambdaComponentArgs:
    """Arguments for Lambda Component."""

    def __init__(
        self,
        environment_suffix: str,
        env_config: EnvironmentConfig,
        role_arn: str,
        tags: Optional[Dict[str, str]] = None
    ):
        self.environment_suffix = environment_suffix
        self.env_config = env_config
        self.role_arn = role_arn
        self.tags = tags or {}


class LambdaComponent(pulumi.ComponentResource):
    """
    Reusable Lambda component for payment processing.
    Creates Lambda function with ARM64 architecture and environment-specific settings.
    """

    def __init__(
        self,
        name: str,
        args: LambdaComponentArgs,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('payment:compute:LambdaComponent', name, None, opts)

        child_opts = ResourceOptions(parent=self)

        # Create Lambda function code
        # In production, this would be actual payment processing code
        lambda_code = '''
import json
import os

def handler(event, context):
    """Process payment transaction."""
    environment = os.environ.get('ENVIRONMENT', 'unknown')

    # Simulate payment processing
    transaction_id = event.get('transaction_id', 'unknown')
    amount = event.get('amount', 0)

    print(f"Processing payment in {environment} environment")
    print(f"Transaction ID: {transaction_id}, Amount: ${amount}")

    return {
        'statusCode': 200,
        'body': json.dumps({
            'message': 'Payment processed successfully',
            'transaction_id': transaction_id,
            'environment': environment,
            'amount': amount
        })
    }
'''

        # Create Lambda function
        self.function = aws.lambda_.Function(
            f"payment-processor-{args.environment_suffix}",
            runtime="python3.11",
            role=args.role_arn,
            handler="index.handler",
            code=pulumi.AssetArchive({
                'index.py': pulumi.StringAsset(lambda_code)
            }),
            memory_size=args.env_config.lambda_memory_mb,
            timeout=30,
            architectures=["arm64"],  # ARM64 for cost optimization
            environment=aws.lambda_.FunctionEnvironmentArgs(
                variables={
                    'ENVIRONMENT': args.env_config.name,
                    'LOG_LEVEL': 'INFO',
                }
            ),
            tags={
                **args.tags,
                'Name': f"payment-processor-{args.environment_suffix}",
                'Architecture': 'arm64',
            },
            opts=child_opts
        )

        # Create CloudWatch Log Group
        self.log_group = aws.cloudwatch.LogGroup(
            f"payment-processor-logs-{args.environment_suffix}",
            name=self.function.name.apply(lambda name: f"/aws/lambda/{name}"),
            retention_in_days=args.env_config.s3_log_retention_days,
            tags=args.tags,
            opts=child_opts
        )

        # Register outputs
        self.register_outputs({
            'function_arn': self.function.arn,
            'function_name': self.function.name,
            'log_group_name': self.log_group.name,
        })
